src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ModelName/SuffixAnalyzerBase.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Azure.ClientSdk.Analyzers.ModelName
+{
+    /// <summary>
+    /// Base analyzer to check model name suffixes.
+    /// </summary>
+    public abstract class SuffixAnalyzerBase : DiagnosticAnalyzer
+    {
+        protected static readonly string Title = "Improper model name suffix";
+        protected static readonly string Description = "Suffix is not recommended. Consider to remove or modify it.";
+        protected static readonly string GeneralRenamingMessageFormat = "Model name '{0}' ends with '{1}'. Suggest to rename it to an appropriate name.";
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray<DiagnosticDescriptor>.Empty;
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.EnableConcurrentExecution();
+            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            var typeSymbol = (INamedTypeSymbol)context.Symbol;
+            if (!IsClassUnderModelsNamespace(typeSymbol) || AnalyzerUtils.IsNotSdkCode(typeSymbol))
+                return;
+
+            var match = SuffixRegex.Match(typeSymbol.Name);
+            if (match.Success && !ShouldSkip(typeSymbol, context))
+            {
+                var suffix = match.Groups["Suffix"].Value;
+                context.ReportDiagnostic(GetDiagnostic(typeSymbol, suffix, context));
+            }
+        }
+
+        protected virtual bool ShouldSkip(INamedTypeSymbol symbol, SymbolAnalysisContext context) => false;
+        protected virtual Regex SuffixRegex => throw new NotImplementedException();
```

Comment:
Since these regexes are just doing the equivalent of "EndsWith", it would be more performant to make this something like below 

```suggestion
        protected abstract string[] InvalidSuffixes;
```

Then the implementation of `AnalyzeSymbol` changes to iterate through InvalidSuffixes to see if `typeSymbol.Name` ends with any of them.

I did a quick perf test just to verify that this is indeed faster, even with checking the worst case of the general suffixes.

|   Method |       Mean |     Error |   StdDev | Ratio | Allocated | Alloc Ratio |
|--------- |-----------:|----------:|---------:|------:|----------:|------------:|
| EndsWith |   193.9 ns |   3.76 ns |  4.48 ns |  0.02 |         - |          NA |
|    RegEx | 8,544.4 ns | 100.82 ns | 94.30 ns |  1.00 |         - |          NA |

Reply:
replaced with strings